File: java/TSPSolution.java
import java.util.ArrayList;
import java.util.Arrays;

public class TSPSolution {
    private TSPData data_;
    private ArrayList<Integer> nodes_;

    public TSPSolution(TSPData data, ArrayList<Integer> nodes) {
	data_ = data;
	nodes_ = nodes;
    }
    
    public int two_opt() {
	int t = 0;
	while (first_improvement()) t++;
	return t;
    }

    private boolean first_improvement() {
	for (int p1=0; p1 < nodes_.size() - 3; p1++) {
	    for (int p2=p1+2; p2 < nodes_.size() -1; p2++) {
		if (data_.d(nodes_.get(p1), nodes_.get(p1+1)) +
		    data_.d(nodes_.get(p2), nodes_.get(p2+1)) > 
		    data_.d(nodes_.get(p1), nodes_.get(p2)) +
		    data_.d(nodes_.get(p1+1), nodes_.get(p2+1))) {
		    // improving two-exchange found!
		    int t;
		    for (int i=0; i < (p2-p1+1) / 2; i++) {
			t = nodes_.get(p1+1+i);
			nodes_.set(p1+1+i, nodes_.get(p2-i));
			nodes_.set(p2-i, t);
		    }
		    return true;
		}
	    }
	}
	return false;
    }
}

File: numba-flat-matrix/tspsolution.py
import itertools as it
import numpy as np
from numba import jit

from tspdata import dist

class TSPSolution:
    def __init__(self, data, permutation):
        self.data = data
        self.nodes = np.array(permutation)

    def two_opt(self):
        t = 0
        while self.first2eimprovement():
            t += 1
        return t

    def first2eimprovement(self):
        # actually speeds things up!
        tour, d, n = self.nodes, self.data.d, self.data.n
        return aux_2opt(tour, d, n)

    def or_opt(self):
        t = 0
        while self.firstorimprovement():
            t += 1
        return t
        
    def firstorimprovement(self):
        # actually speeds things up!
        tour, d, n = self.nodes, self.data.d, self.data.n
        return aux_oropt(tour, d, n)
    
@jit(nopython=True)
def aux_2opt(tour, d, n):
    for p1 in range(len(tour) - 3):
        for p2 in range(p1+2, len(tour) - 1):
            if dist(d, n, tour[p1], tour[p1+1]) + dist(d, n, tour[p2], tour[p2+1]) > \
               dist(d, n, tour[p1], tour[p2]) + dist(d, n, tour[p1+1], tour[p2+1]):
                # improving 2-exchange found
                for i in range((p2-p1+1) // 2):
                    tour[p1+1+i], tour[p2-i] = tour[p2-i], tour[p1+1+i]
                return True
    return False

@jit(nopython=True)
def aux_oropt(tour, d, n):
    for i in range(1, len(tour) - 1):
        for l in range(1, 1 + min(3, len(tour)-1-i)):
            for tmp in (range(i-1), range(i+l, len(tour)-1)):
                for p in tmp:
                    delta = dist(d, n, tour[i-1], tour[i+l]) + dist(d, n, tour[p], tour[i]) + \
                        dist(d, n, tour[i+l-1], tour[p+1]) - dist(d, n, tour[p], tour[p+1])\
                        - dist(d, n, tour[i-1], tour[i]) - dist(d, n, tour[i+l-1], tour[i+l])
                    # perform improving move
                    if delta < 0:
                        # store sequence to move
                        t = np.array([ x for x in tour[i:i+l] ])
                        if i < p:
                            # shift stuff left
                            tour[i:p-l+1] = tour[i+l:p+1]
                            # copy sequence right of the stuff
                            tour[p+1-l:p+1] = t[:]
                        else:
                            # shift stuff right
                            tour[i-1+l:p+l:-1] = tour[i-1:p:-1]
                            # copy sequence right of the stuff
                            tour[p+1:p+1+l] = t[:]
                        return True
    return False
